src/box/tx_memory.h

```diff
@@ -0,0 +1,212 @@
+ * @a tx_stat, create instance of derivated memory manager instead.
+ */
+
+#include <trivia/util.h>
+#include "small/mempool.h"
+#include "histogram.h"
+struct txn;
+struct tuple;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Allocations types for txns.
+ */
+enum TXN_ALLOC_TYPE {
+	TXN_ALLOC_STMT = 0,
+	TXN_ALLOC_SVP = 1,
+	TXN_ALLOC_USER_DATA = 2,
+	TXN_ALLOC_REDO_LOG = 3,
+	TXN_ALLOC_TRIGGER = 4,
+	TXN_ALLOC_TIMER = 5,
+	TXN_ALLOC_JOURNAL_ENTRY = 6,
+	TXN_ALLOC_MAX = 7
+};
+
+/**
+ * String appearance of enum TXN_ALLOC_TYPE
+ */
+extern const char *TXN_ALLOC_TYPE_STRS[];
+
+/**
+ * Storage for statistics of one allocation type for txn.
+ */
+struct txn_stat_storage {
+	struct histogram *hist;
+	uint64_t total;
+};
+
+/**
+ * Memory manager itself.
+ */
+struct tx_memory_manager {
+	/** Number of allocation types. */
+	uint32_t alloc_max;
+	/** Number of registered txn. */
+	uint64_t txn_num;
+	/**
+	 * Pointer to array of @txn_stat_storage objects.
+	 * It is better to allocate it as a field of derived memory manager.
+	 */
+	struct txn_stat_storage *stats_storage;
+};
+
+/**
+ * Memory usage of one txn. Allocated on its region.
+ */
+struct txn_mem_used {
+	/**
+	 * This field is used only in debug mode to make sure that txn has
+	 * deallocated all the mempool allocations before it was deleted.
+	 */
+	uint64_t mempool_total;
+	/** Total memory used for every type of allocations. */
+	uint64_t total[];
+};
+
+/**
+ * @brief A wrapper over region_alloc.
+ * @param txn Owner of a region.
+ * @param size Bytes to allocate.
+ * @param alloc_type See TXN_ALLOC_TYPE.
+ * @note The only way to truncate a region of @a txn is to clear @a txn.
+ */
+void *
+tx_memory_region_alloc(struct tx_memory_manager *stat, struct txn *txn,
+		       size_t size, size_t alloc_type);
+
+/**
+ * @brief Register @a txn in @a tx_stat. It is very important
+ * to register txn before using allocators from @a tx_stat.
+ */
+void
+tx_memory_register_txn(struct tx_memory_manager *stat, struct txn *txn);
+
+/**
+ * @brief Unregister @a txn and truncate its region up to sizeof(txn).
+ */
+void
+tx_memory_clear_txn(struct tx_memory_manager *stat, struct txn *txn);
+
+/**
+ * @brief A wrapper over region_aligned_alloc.
+ * @param txn Owner of a region.
+ * @param size Bytes to allocate.
+ * @param alignment Alignment of allocation.
+ * @param alloc_type Type of allocation.
+ */
+void *
+tx_memory_region_aligned_alloc(struct tx_memory_manager *stat, struct txn *txn,
+			size_t size, size_t alignment, size_t alloc_type);
+
+#define tx_memory_region_alloc_object(stat, txn, T, size, alloc_type) ({  \
+	*(size) = sizeof(T);                                              \
+	(T *)tx_memory_region_aligned_alloc((stat), (txn), sizeof(T),     \
+	                                    alignof(T), (alloc_type));    \
+})
+
+/**
+ * @brief Getter for txn's region. Use only if region was not given before.
```

Comment:
The purpose of give/take methods is very unclear. I'd add an explanation how you are going to utilize them